### ISSUE TITLE
Reading a datatype that is set to null returns a default value instead of giving you back an error

### ISSUE THREAD

**Issue by calvinlfer** [ringcentral/cassandra4io#12]
Hey there,

I have found that reading values that could potentially be null but reading them as a non-optional datatype will return default values instead of providing an error. I believe this is the default behavior of the underlying Datastax Java client.

Here's an example:
```cql
CREATE TABLE example_table(
    a int,
    b ascii,
    c double,
    PRIMARY KEY(a)
);

INSERT INTO example_table(a) VALUES (1);
```

Reading this using the following:
```scala
final case class TableRow(a: Int, b: String, c: Double)

class TableRepo(session: CassandraSession[IO]) {
    def get(a: Int): Stream[IO, TableRow] =
      cql"SELECT a, b, c FROM example_table WHERE a = $a"
        .as[TableRow]
        .select(session)
  }
```

Doing that will give you back a `TableRow(a = 1, b = null, c = 0.0)`
I was wondering if we could impose an additional constraint about fields not being null and if they were null then throwing an error (rather than returning default values that could be confusing - for example here double returns 0.0 and boolean returns false when null) when you tried to read these fields from Cassandra. I have managed to do this by doing the following:

```scala
package com.experiments.calvin
import com.datastax.oss.driver.api.core.cql.{BoundStatement, Row}
import com.ringcentral.cassandra4io.cql.{Binder, Reads}
import shapeless.<:!<

final case class NonNullableConstraintViolated(message: String) extends Exception

final case class NonNullable[A](underlying: A)
object NonNullable {
  implicit def nonNullableCqlReads[A](implicit
    underlyingReads: Reads[A],
    evidenceNotNested: A <:!< NonNullable[_]
  ): Reads[NonNullable[A]] = { (row: Row, index: Int) =>
    if (row.isNull(index))
      throw NonNullableConstraintViolated(
        s"Expected valid data at index $index at row ${row.getFormattedContents} but got NULL instead"
      )
    else {
      val (underlying, nextIndex) = underlyingReads.read(row, index)
      (NonNullable(underlying), nextIndex)
    }
  }

  implicit def nonNullableCqlBinder[A](implicit
    underlyingBinder: Binder[A],
    evidenceNotOption: A <:!< Option[_],
    evidenceNotNested: A <:!< NonNullable[_]
  ): Binder[NonNullable[A]] =
    (statement: BoundStatement, index: Int, value: NonNullable[A]) =>
      underlyingBinder.bind(statement, index, value.underlying)
}
```

There are some extra implicit requirements to prevent doing thing like `NonNullable[Option[A]]` or `NonNullable[NonNullable[A]]`. I was wondering if you consider adding a mechanism/constraint like this to the library that would now throw errors if you read data that you did not expect to be null or would you happen to know if you can tweak this behavior within the core Datastax Java client itself?

Thank you for reading
Cal

**Comment by Alexey-Yuferov** [comment]
Hello, @calvinlfer !
Thank you for your input. 
Indeed this behaviour is not desirable and is inhered from Datastax: https://docs.datastax.com/en/drivers/java/4.13/com/datastax/oss/driver/api/core/data/GettableByIndex.html#getInt-int- for example.
However, I am not sure that throwing runtime exception is a good solution. For instance, you could use Option in your case class, i.e. if. you anyway delegate this to data model why not just state optionality explicitly and simply?
@narma , what do you think?

**Comment by calvinlfer** [comment]
Hey @Alexey-Yuferov thats a good point. We had a use case at work where there was a mismatch in expectation of the Cassandra data model (when migrating a legacy application over) since the Cassandra data types don’t explicitly signal a value can be null and when you read out the data, you’ll end up reading out default values when you read nulls (like for example false when you encounter a Boolean null) which can lead to subtle bugs in application logic which is why I was suggesting this more rigorous form of reading data, I know it’s not a common use case but I just wanted to see what you guys thought

**Comment by calvinlfer** [comment]
Yeah that’s fair, the NonNullable type was just an aggressive check to prevent subtle bugs when reading values that you don’t expect to be null but you are right - a user can easily incorporate this feature

**Comment by narma** [comment]
I would like to reopen this issue, In term of safety library should provide guarantee that null values will be accepted only for Option[T]
But I will suggest different approach to resolve it without involving new typeclass in new PR.
